src/UIA.cpp

```diff
@@ -0,0 +1,207 @@
+/*
+ * OSARA: Open Source Accessibility for the REAPER Application
+ * UIAutomation code
+ * Author: James Teh <jamie@jantrid.net>
+ * Copyright 2019 NV Access Limited, James Teh
+ * License: GNU General Public License version 2.0
+ */
+
+#include "osara.h"
+#include <UIAutomation.h>
+#include <ole2.h>
+
+using namespace std;
+
+HWND UIAWnd = 0;
+const char* WINDOW_CLASS_NAME = "REAPEROSARANotificationWND";
+typedef HRESULT(WINAPI *UiaRaiseNotificationEvent_funcType)(
+	_In_ IRawElementProviderSimple* provider,
+	NotificationKind notificationKind,
+	NotificationProcessing notificationProcessing,
+	_In_opt_ BSTR displayString,
+	_In_ BSTR activityId
+);
+HMODULE UIAutomationCore = nullptr;
+UiaRaiseNotificationEvent_funcType UiaRaiseNotificationEvent_ptr = nullptr;
+
+// Provider code based on Microsoft's UIAutomationSimpleProvider example.
+class UIAProviderImpl : public IRawElementProviderSimple
+{
+public:
+	UIAProviderImpl(_In_ HWND hwnd): controlHWnd(hwnd) {}
+
+	// IUnknown methods
+	ULONG STDMETHODCALLTYPE AddRef() {
+		return InterlockedIncrement(&refCount);
+	}
+
+	ULONG STDMETHODCALLTYPE Release() {
+		long val = InterlockedDecrement(&refCount);
+		if (val <= 0) {
+			delete this;
+		}
+		return val;
+	}
+
+	HRESULT STDMETHODCALLTYPE QueryInterface(_In_ REFIID riid, _Outptr_ void** ppInterface) {
+		if (ppInterface) {
+			return E_INVALIDARG;
+		}
+		if (riid == __uuidof(IUnknown)) {
+			// Several examples seem to cast to IRawElementProviderSimple instead of IUnknown in this case.
```

Comment:
This is done because in C++, where there is multiple inheritance, there can be multiple IUnknowns. All COM interfaces derive from IUnknown. So, for example, if you inherit both IRawElementProviderSimple and IRawElementProviderFragment, there are two different IUnknown pointers. COM requires that when you query to IUnknown, you should always get the same pointer, regardless of the originating interface. Casting to IRawElementProviderSimple first ensures you always get the same IUnknown.

Because you're only inheriting from one interface, this technically isn't necessary, but it's good practice because it's very easy to forget to make this change if you later add an additional interface.